### PR TITLE
Fix not removed watches on unallowed repositories

### PR DESCRIPTION
**Pull request by daviian**
Targets: https://github.com/go-gitea/gitea/issues/3343, https://github.com/go-gitea/gitea/issues/3782, https://github.com/go-gitea/gitea/issues/4149

What does it do:
* removes watches if users gets removed from a team or a team is deleted as a whole.
* also sets new team users to watch all team repositories.
* remove watch if user gets removed from the collaborators of a repo

TODO:
- [x] create migration to remove watches that are hanging in the air
- [x] also remove issue_watches
- [x] also remove issue_watches in migration

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

**Comment by daviian** [comment]
@lafriks Please add status/wip as I've to add the migration to cleanup hanging watches

**Comment by bkcsoft** [comment]
@daviian I don't think that the mgiration is necessary for this fix. It would be nice to have, but not something that should hold back this PR.

**Comment by daviian** [comment]
@bkcsoft  It's already implemented. I'm currently in the middle of testing the migration locally ;) If it's fine i'll push.

File: models/migrations/v67.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		}
+
+		a := &Access{UserID: userID, RepoID: repo.ID}
+		if has, err := x.Get(a); !has || err != nil {
+			return mode, err
+		}
+		return a.Mode, nil
+	}
+
+	sess := x.NewSession()
```
Comment (lunny) [inline models/migrations/v67.go:67]: need 
```Go
defer sess.Close()
if err := sess.Begin(); err != nil {
return err
}
```

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/org_team.go [hunk 2]
```diff
@@ -178,6 +178,11 @@ func (t *Team) removeRepository(e Engine, repo *Repository, recalculate bool) (e
 		if err = watchRepo(e, teamUser.UID, repo.ID, false); err != nil {
 			return err
 		}
+
+		// Remove all IssueWatches a user has subscribed to in the repositories
+		if err := removeIssueWatchersByRepoID(e, teamUser.ID, repo.ID); err != nil {
```
Comment (lunny) [inline models/org_team.go]: teamUser.UID

File: models/migrations/v67.go (hunk 1 above)
Comment (daviian) [reply models/migrations/v67.go:67]: Oh yes. Totally forgot.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by daviian** [comment]
@lunny Added your suggestions

File: models/migrations/v67.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			watch := bean.(*Watch)
+
+			repo := repoCache[watch.RepoID]
+			if repo == nil {
+				repo = &Repository{
+					ID: watch.RepoID,
+				}
+				if _, err := x.Get(repo); err != nil {
+					return err
+				}
```
Comment (lunny) [inline models/migrations/v67.go:85]: You should `repoCache[watch.RepoID] = repo`

File: models/migrations/v67.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+				watch := bean.(*IssueWatch)
+
+				repo := repoCache[watch.RepoID]
+				if repo == nil {
+					repo = &Repository{
+						ID: watch.RepoID,
+					}
+					if _, err := x.Get(repo); err != nil {
+						return err
+					}
```
Comment (lunny) [inline models/migrations/v67.go:126]: The same as above

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
Please follow my comment, otherwise LGTM

**Review by lunny — APPROVED** [review]

**Comment by techknowlogick** [comment]
@lafriks the ticket this was attached to was marked as backport, but you removed the label from this PR, should it still be backported?

**Comment by lunny** [comment]
This will not be back port since it has a migration which will let user difficult upgrade from v1.4 to v1.5

**Pull request merged**